packages/next/src/server/after/react-cache-scope.ts

```diff
@@ -0,0 +1,188 @@
+import { AsyncLocalStorage } from 'async_hooks'
+
+export function createCacheScope() {
+  const storage = createCacheMap()
+  return {
+    run: <T>(callback: () => T): T => {
+      return CacheScopeStorage.run(storage, () => callback())
+    },
+  }
+}
+
+export type CacheScope = ReturnType<typeof createCacheScope>
+
+// #region  custom cache dispatcher with support for scoping
+
+// Note that the plan is to upstream this into React itself,
+// but after() is an experimental feature, and this should be good enough for now.
+
+type CacheDispatcher = {
+  [HAS_CACHE_SCOPE]?: boolean
+  getCacheForType: <T>(create: () => T) => T
+  // DEV-only (or !disableStringRefs)
+  getOwner?: () => null | Record<string, unknown>
+}
+
+const HAS_CACHE_SCOPE: unique symbol = Symbol.for('next.cacheScope')
+
+type CacheMap = Map<Function, unknown>
+
+function createCacheMap(): CacheMap {
+  return new Map()
+}
+
+function isWithinCacheScope() {
+  return !!CacheScopeStorage.getStore()
+}
+
+const CacheScopeStorage: AsyncLocalStorage<CacheMap> =
+  new AsyncLocalStorage<CacheMap>()
+
+/** forked from packages/react-server/src/flight/ReactFlightServerCache.js */
+function resolveCache(): CacheMap {
+  const store = CacheScopeStorage.getStore()
+  if (store) {
+    return store
+  }
+  return createCacheMap()
+}
+
+/** forked from packages/react-server/src/flight/ReactFlightServerCache.js */
+const ScopedCacheDispatcher: CacheDispatcher = {
+  getCacheForType<T>(resourceType: () => T): T {
+    if (!isWithinCacheScope()) {
+      throw new Error(
+        'Invariant: Expected patched cache dispatcher to run within CacheScopeStorage'
+      )
+    }
+    const cache = resolveCache()
+    let entry: T | undefined = cache.get(resourceType) as any
+    if (entry === undefined) {
+      entry = resourceType()
+      // TODO: Warn if undefined?
+      cache.set(resourceType, entry)
+    }
+    return entry
+  },
+}
+
+// #endregion
+
+// #region  injecting the patched dispatcher into React
+
+export function patchCacheScopeSupportIntoReact(React: typeof import('react')) {
+  const internals = getReactServerInternals(React)
+  if ('ReactCurrentCache' in internals) {
+    // react < 19
+    const { ReactCurrentCache } = internals
+    if (ReactCurrentCache.current) {
+      patchReactCacheDispatcher(ReactCurrentCache.current)
+    } else {
+      patchReactCacheDispatcherWhenSet(ReactCurrentCache, 'current')
+    }
+  } else if ('A' in internals) {
+    // react >= 19
+    if (internals.A) {
+      patchReactCacheDispatcher(internals.A)
+    } else {
+      patchReactCacheDispatcherWhenSet(internals, 'A')
+    }
+  } else {
+    throw new Error(
+      'Invariant: Could not find cache dispatcher in React internals'
```

Comment:
Not a fan of these "invariant" prefixes since it doesn't tell users what's up. Let's just make it explicit in the error message that this would be a Next.js bug if encountered.

Reply:
Any idea how to phrase this in a more helpful way? We have a bunch of messages like `"XYZ. This is a bug in Next.js"`, so I guess maybe we do that instead of `"Invariant: XYZ"`?

Reply:
Yep. That's also what we do in React. Helps more with figuring out what to do this error: file a bug report.

Reply:
Added an `InvariantError` that templates the message as `"Invariant {message}. This is a bug in Next.js."`